signatures/pdf_sign_external_globalsign.go

```diff
@@ -0,0 +1,539 @@
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"time"
+
+	"github.com/unidoc/globalsign-dss"
+	"github.com/unidoc/pkcs7"
+	"github.com/unidoc/unipdf/v3/annotator"
+	"github.com/unidoc/unipdf/v3/common/license"
+	"github.com/unidoc/unipdf/v3/core"
+	"github.com/unidoc/unipdf/v3/model"
+
+	"golang.org/x/crypto/ocsp"
+)
+
+func init() {
+	// Make sure to load your metered License API key prior to using the library.
+	// If you need a key, you can sign up and create a free one at https://cloud.unidoc.io
+	err := license.SetMeteredKey(os.Getenv(`UNIDOC_LICENSE_API_KEY`))
+	if err != nil {
+		panic(err)
+	}
+}
+
+var (
+	sigLen = 8192
+
+	apiKey, apiSecret, certFilepath, keyFilepath string
+)
+
+const usagef = "Usage: %s INPUT_PDF_PATH OUTPUT_PDF_PATH API_KEY API_SECRET CERT_FILE_PATH KEY_FILE_PATH\n"
+
+func main() {
+	args := os.Args
+	if len(args) < 7 {
+		fmt.Printf(usagef, os.Args[0])
+		return
+	}
+	inputPath := args[1]
+	outputPath := args[2]
+
+	apiKey = args[3]
+	apiSecret = args[4]
+	certFilepath = args[5]
+	keyFilepath = args[6]
+
+	// This would be the time to send the PDF buffer to a signing device or
+	// signing web service and get back the signature. We will simulate this by
+	// signing the PDF using UniDoc and returning the signature data.
+	pdfData, _, err := getExternalSignatureAndSign(inputPath)
+	if err != nil {
+		log.Fatalf("Fail signature: %v\n", err)
+	}
+
+	// Write output file.
+	if err := ioutil.WriteFile(outputPath, pdfData, os.ModePerm); err != nil {
+		log.Fatalf("Fail: %v\n", err)
+	}
+
+	log.Printf("PDF file successfully signed. Output path: %s\n", outputPath)
+}
+
+// generateSignedFile generates a signed version of the input PDF file using the
+// specified signature handler.
+func generateSignedFile(inputPath string, handler model.SignatureHandler, field *model.PdfFieldSignature) ([]byte, error) {
+	// Create reader.
+	file, err := os.Open(inputPath)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	reader, err := model.NewPdfReader(file)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create appender.
+	appender, err := model.NewPdfAppender(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = appender.Sign(1, field); err != nil {
+		return nil, err
+	}
+
+	// Add LTV into signature.
+	var certChain []*x509.Certificate
+	if getter, ok := handler.(*externalSigner); ok {
+		certChain = getter.getCertificateChain()
+	}
+
+	ltv, err := model.NewLTV(appender)
+	if err != nil {
+		return nil, err
+	}
+	ltv.CertClient.HTTPClient.Timeout = 30 * time.Second
+	ltv.OCSPClient.HTTPClient.Timeout = 30 * time.Second
+	ltv.CRLClient.HTTPClient.Timeout = 30 * time.Second
+
+	err = ltv.EnableChain(certChain)
+	if err != nil {
+		return nil, err
+	}
+
+	// Write PDF file to buffer.
+	pdfBuf := bytes.NewBuffer(nil)
+	if err = appender.Write(pdfBuf); err != nil {
+		return nil, err
+	}
+
+	return pdfBuf.Bytes(), nil
+}
+
+// getExternalSignature simulates an external service which signs the specified
```

Comment:
The name of the function differs from the one mentioned in the comment as well: 
`getExternalSignature` -> `getExternalSignatureAndSign`.
Also, the comment should be changed a bit as the external service is not simulated.

Reply:
Updated at 15bb7144b66b8891f5c68bd64803bb2d9ee32f54